### PR TITLE
Add sparse `gather` `scatter` `all_gather`

### PR DESCRIPTION
**Pull request by emcastillo**
Blocked by #6762

### CODE REVIEW

File: cupyx/distributed/_nccl_comm.py [hunk 1]
```diff
@@ -485,12 +515,20 @@ def _exchange_shape_and_sizes(
...
                 sizes_shape = numpy.empty(5, dtype='q')
                 comm._mpi_comm.Recv(sizes_shape, source=peer, tag=1)
                 return sizes_shape
+            elif method == 'bcast':
+                sizes_shape = numpy.array(sizes_shape, dtype='q')
+                return comm._mpi_comm.bcast(sizes_shape, root=peer)
+            elif method == 'gather':
+                sizes_shape = numpy.array(sizes_shape, dtype='q')
+                # recv_buf = numpy.empty([comm._n_devices, 5], dtype='q')
+                return comm._mpi_comm.gather(sizes_shape, peer)
```
Comment (leofang) [inline cupyx/distributed/_nccl_comm.py:530]: Ditto

File: cupyx/distributed/_nccl_comm.py [hunk 2]
```diff
@@ -131,7 +130,10 @@ def _get_op(self, op, dtype):
 
     def _dispatch_arg_type(self, function, args):
         comm_class = _DenseNCCLCommunicator
-        if sparse.issparse(args[0]):
+        if (
+            (isinstance(args[0], list) and sparse.issparse(args[0][0]))
+            or sparse.issparse(args[0])
+        ):
```
Comment (asi1024) [inline cupyx/distributed/_nccl_comm.py:136]: May args be a tuple?

**Comment by emcastillo** [comment]
Review comments addressed, PTAL!

**Review by asi1024 — APPROVED** [review]

**Pull request merged**